### PR TITLE
Add package upgrade list

### PR DESCRIPTION
**Pull request by Doprez**
# PR Details

There are some cases where nugets contain "Stride." but do not follow the same release version. This adds a list that can be used to check if the package should be upgraded.

This is a replacement for https://github.com/stride3d/stride/pull/1881

below is a test I did excluding `Stride.Physics` and `Stride.Navigation` showing that no other packages are effected but the ones not in the array.
image
image

## Description

Adds a static array of valid projects and adds a couple checks for initial upgrades and sequential upgrades.

## Related Issue

https://github.com/stride3d/stride/issues/2053

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] **I have built and run the editor to try this change out.**

### CODE REVIEW

**Comment by Doprez** [comment]
Once this gets reviewed I would like to do a demo video using the community toolkit with stride but this error is mildly annoying for me and Im sure if I advertise it many average users would deal with the same issue.

File: sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Stride.Core.Assets;
+public static class StridePackagesToUpgrade
+{
+    public static string[] PackageNames =
```
Comment (Kryptos-FR) [inline sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs:10]: `Contains()` in a `List<T>` or `Array` is a *O(n)* operation. Better use a `HashSet<string>` for speed.

Comment (Kryptos-FR) [reply sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs:10]: And if possible, sort them by name to make it easier to maintain.

Comment (Doprez) [reply sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs]: good calls, changed to `HashSet` and alphabetized the list per region and cascading.

**Comment by VaclavElias** [comment]
Maybe we should go with this solution "temporarily" (long term) till the Stride reservation is resolved and taken care of.

**Comment by Doprez** [comment]
I think this list will still be useful for some edge cases but you are right that Stride reservation should be put in place. There were still some packages that had Stride but were not on the same version that I found in nuget.org.

https://www.nuget.org/packages/Stride.GraphX.WPF.Controls
https://www.nuget.org/packages/Stride.GNU.Gettext
https://www.nuget.org/packages/Stride.OpenTK
https://www.nuget.org/packages/Stride.Metrics

and a few more but these are some of the examples that may cause issues for adventerous nuget users.

File: sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
+using System;
```
Comment (Kryptos-FR) [inline sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs:1]: Nit-pick: add the header.

```csharp
// Copyright (c) .NET Foundation and Contributors (https://dotnetfoundation.org/ & https://stride3d.net)
// Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
```

File: sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Stride.Core.Assets;
+public static class StridePackagesToUpgrade
```
Comment (Kryptos-FR) [inline sources/assets/Stride.Core.Assets/StridePackagesToUpgrade.cs:8]: Nit-pick: separate by an empty line.

```suggestion
namespace Stride.Core.Assets;

public static class StridePackagesToUpgrade
```

**Review by Kryptos-FR — CHANGES_REQUESTED** [review]

**Comment by Eideren** [comment]
Xen2 sent a request for reserving the prefix over on nuget so I think it's safe to change from a whitelist to a blacklist instead; list only the specific repo under the prefix that should not be updated instead of listing those that should be. So that would be `Stride.Community.*`, `Stride.Awesome.Shaders` and the other you listed above

**Comment by Doprez** [comment]
> Xen2 sent a request for reserving the prefix over on nuget so I think it's safe to change from a whitelist to a blacklist instead; list only the specific repo under the prefix that should not be updated instead of listing those that should be. So that would be `Stride.Community.*`, `Stride.Awesome.Shaders` and the other you listed above

I changed over to an exclusion list like you said.

here is the quick test I did with the changes
exclusion list

image

result

image

**Comment by Doprez** [comment]
@Eideren I addded a for loop to check the `starts with` validation as well. I couldn't quickly think of a better way than using `ElementAt()` since its a hashset and not a list I cant just index normally so I feel like this defeats the purpose of a hashset but I could be wrong.

**Comment by Eideren** [comment]
Sorry about the delay. Yep, just swap to an array, do an ordinal string equality and we're done :)

**Comment by Doprez** [comment]
Not a problem at all, I think Im done with this PR now.

**Comment by Eideren** [comment]
Thanks !

**Pull request merged**